services/impl/src/main/java/org/exoplatform/automatic/translation/impl/connectors/GoogleTranslateConnector.java

```diff
@@ -16,12 +16,96 @@
+import org.apache.http.entity.StringEntity;
+import org.apache.http.impl.DefaultConnectionReuseStrategy;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.impl.conn.PoolingHttpClientConnectionManager;
 import org.exoplatform.automatic.translation.api.AutomaticTranslationComponentPlugin;
 import org.exoplatform.commons.api.settings.SettingService;
+import org.exoplatform.services.log.ExoLogger;
+import org.exoplatform.services.log.Log;
+import org.json.JSONObject;
+
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.nio.charset.StandardCharsets;
+import java.util.Locale;
 
 public class GoogleTranslateConnector extends AutomaticTranslationComponentPlugin {
 
+  private static final Log    LOG                      = ExoLogger.getLogger(GoogleTranslateConnector.class);
+
+  private static final String GOOGLE_TRANSLATE_SERVICE = "google-translate";
+
+  private static final String API_URL                  = "https://translation.googleapis.com/language/translate/v2";
+
+  private static final String KEY_PARAM                = "key";
+
+  private static final String DATA_PATTERN             = "{'q': '{message}', 'target': '{targetLocale}'}";
+
+  private static final int DEFAULT_POOL_CONNECTION = 100;
+
+  private HttpClient httpClient;
+
   public GoogleTranslateConnector(SettingService settingService) {
     super(settingService);
+
+
+    PoolingHttpClientConnectionManager connectionManager = new PoolingHttpClientConnectionManager();
+    connectionManager.setDefaultMaxPerRoute(DEFAULT_POOL_CONNECTION);
+    HttpClientBuilder httpClientBuilder = HttpClients.custom()
+                                                     .setConnectionManager(connectionManager)
+                                                     .setConnectionReuseStrategy(new DefaultConnectionReuseStrategy())
+                                                     .setMaxConnPerRoute(DEFAULT_POOL_CONNECTION);
+    this.httpClient = httpClientBuilder.build();
   }
+
+  @Override
+  public String translate(String message, Locale targetLocale) {
+    long startTime = System.currentTimeMillis();
+
+    String serviceUrl = API_URL + "?" + KEY_PARAM + "=" + getApiKey();
+    String data = DATA_PATTERN.replace("{message}", message).replace("{targetLocale}", targetLocale.getLanguage());
+    try {
+      HttpPost httpTypeRequest = new HttpPost(serviceUrl);
+      httpTypeRequest.setEntity(new StringEntity(data, ContentType.APPLICATION_JSON));
+      HttpResponse httpResponse=httpClient.execute(httpTypeRequest);
+      String response = null;
+      int statusCode = httpResponse.getStatusLine().getStatusCode();
+      if (statusCode == HttpURLConnection.HTTP_OK) {
+
+        // read the response
+        if (httpResponse.getEntity()!=null) {
+          try (InputStream is = httpResponse.getEntity().getContent()){
+            response = IOUtils.toString(is, StandardCharsets.UTF_8);
```

Comment:
please close the input stream

Reply:
I follow the link you gave me yesterday https://www.javamadesoeasy.com/2015/05/try-with-resources-in-java.html
So I think this is ok

Reply:
It's ok